scipy/optimize/zeros.py

```diff
@@ -209,6 +211,93 @@ def newton(func, x0, fprime=None, args=(), tol=1.48e-8, maxiter=50,
     raise RuntimeError(msg)
 
 
+def _array_newton(func, x0, fprime, args, tol, maxiter, fprime2,
```

Comment:
Here comes a pretty big refactoring suggestion... If I understand things correctly, `func`, `fprime` and `fprime2` are expected to be vectorized implementations of a 1D scalar function, right? So they should work regardless of the shape of their input, right?

If those assumptions are right, there probably is some speed to gain by only computing the function for those values for which it is really needed, using a boolean mask. This would also take care of the spurious division by zero warnings now happening. Consider this alternate implementation of the Newton-Raphson branch, untested and probably plagued with errors:

```
x = np.asarray(x0)
active = np.ones_like(x, dtype=bool)
zero_derivative = np.zeros_like(x, dtype=bool)

for iteration in range(maxiter):
    fder = np.asarray(fprime(x[active], *args))
    # Remove zero derivative entries from active set.
    mask = (fder != 0)
    zero_derivative[active] |= ~mask
    active[active] &= mask
    fder = fder[mask]
    if not np.any(active):
        break
    fval = np.asarray(func(x[active], *args))
    newton_step = fval / fder
    if fprime2 is None:
        deltax = newton_step
    else:
        fder2 = np.asarray(fprime2(x[active], *args))
        deltax = newton_step / (1.0 - 0.5 * newton_step * fder2 / fder)
    x[active] -= deltax
    # Remove converged entries from active set.
    active[active] &= np.abs(deltax) >= tol
    if not np.any(active):
        break
if np.any(zero_derivative):
    all_or_some = 'all' if np.all(zero_derivative) else 'some'
    warnings.warn(all_or_some + ' derivatives were zero', RuntimeWarning)
if failure_idx_flags:
    return x, ~active, zero_derivative
return x
```

I have also tried to apply some of my other suggestions when refactoring your code, but the most relevant thing is the use and update of `active` to limit the number of points at which functions are evaluated.

Reply:
This looks great, I'll benchmark it and hopefully there'll be even more performance improvements for arrays by only calculating active elements, and I think it is more readable. Thanks!

Reply:
@jaimefrio there is a problem, this assumes that the only array is `x` but in the [`test_newton_array`](https://github.com/mikofski/scipy/blob/397da36857fdd06e9cb579448f00eddeae5dfeef/scipy/optimize/tests/test_zeros.py#L75-L81) function, some of the extra arguments are also arrays.

Unfortunately, I can't figure out how to know which args are arrays, which are used to calculate the residual and which are used for some private unrelated use, like [this](https://gist.github.com/person142/9789b09c599bda1c2377d496787d4ea0).

We could ask the user, in a separate argument, the args indices that are arrays. I don't think we should make them all arrays, no. I also don't think we should check each arg to see if it's scalar or not, no, definitely not, b/c if the user has a private arg not used for residual, or just used with indices internal to the objective function, then args is jagged and testing each is slow and a inconclusive.

Reply:
That's a bummer... It also seems that your timings don't support this approach as much, although I think we could work around that. Let's leave this out for the moment. If we ever get back to this and see that an approach like this would have a significant impact we can always add an extra boolean flag, something like `clip_inputs`.